### ISSUE TITLE
Enum property hint doesn't support 64-bit integers

### ISSUE THREAD

**Issue by neikeq** [godotengine/godot#22004]
**Godot version:**

5e57beebb1a43c65d949fdfe5081f9c27f181c80

**Issue description:**

The value of exported fields with `PROPERTY_HINT_ENUM` cannot be that of a 64-bit integer.
This is only the case when the field is exported. Everything seems to work well if the field is not exported.

The PR that added support for exporting enums is #12952

**Steps to reproduce:**

Example script:

``` GDScript
const INT_MAX = 2147483647
const UINT_MAX = 4294967295
const LONG_MAX = 9223372036854775807

enum MyEnum {
	A = 0,
	B = 1,
	C = 2147483647,
	D = 2147483648,
	E = 6294967295
}

export(MyEnum) var myvar = E

func _ready():
	print("INT_MAX: ", INT_MAX)
	print("UINT_MAX: ", UINT_MAX)
	print("LONG_MAX: ", LONG_MAX)
	print()
	
	print("myvar: ", myvar)
	print("myvar > INT_MAX: ", myvar > INT_MAX)
	print("myvar > UINT_MAX: ", myvar > UINT_MAX)
	print("myvar > LONG_MAX: ", myvar > LONG_MAX)
	print()
	
	print("C: ", C)
	print("D: ", D)
	print("E: ", E)
	print("C > INT_MAX: ", C > INT_MAX)
	print("D > UINT_MAX: ", D > UINT_MAX)
	print("E > LONG_MAX: ", E > LONG_MAX)
	print()

```

### Output for the constants:

```
INT_MAX: 2147483647
UINT_MAX: 4294967295
LONG_MAX: 9223372036854775807

C: 2147483647
D: 2147483648
E: 6294967295
C > INT_MAX: False
C == INT_MAX: True
C > UINT_MAX: False
C > LONG_MAX: False
D > INT_MAX: True
D > UINT_MAX: False
D > LONG_MAX: False
E > INT_MAX: True
E > UINT_MAX: True
E > LONG_MAX: False
```

- Constants print the correct value.
- `C` comparisons print the expected result.
- `D` comparisons print the expected result, `D > INT_MAX` be `True`.
- `E` comparisons print the expected result, Both `E > INT_MAX` and `E > UINT_MAX` are `True`.

### Output for `myvar = C`:

```
myvar: 2147483647
myvar > INT_MAX: False
myvar == INT_MAX: True
myvar > UINT_MAX: False
myvar > LONG_MAX: False
```

- Prints the correct value.
- Comparisons print the expected result. `myvar == INT_MAX` be `True`.

### Output for `myvar = D`:

```
myvar: -2147483648
myvar > INT_MAX: False
myvar == INT_MAX: False
myvar > UINT_MAX: False
myvar > LONG_MAX: False
```

- Prints the incorrect value: `-2147483648` instead of `2147483648`. **This does not happen if the field is not exported**. It prints the value as a signed integer. This looks weird to me since, even though `Variant::operator String()` uses `itos` which is `String::num_int64` instead of `String::num_uint64`, it operates on a 64-bit signed integer, not a 32-bit signed integer.
- Comparisons DON'T print the expected result. `myvar > INT_MAX` should be `True`, but it's not since it seems to be comparing `-2147483648 > 2147483647`.

### Output for `myvar = E`:

```
myvar: 1999999999
myvar > INT_MAX: False
myvar == INT_MAX: False
myvar > UINT_MAX: False
myvar > LONG_MAX: False
```

- Prints an incorrect value: `1999999999` instead of `6294967295`.
- Comparisons DON'T print the expected result. Both `myvar > INT_MAX` and `myvar > UINT_MAX` should be `True`, but it's not since it seems to be comparing `1999999999 > 2147483647` and `1999999999 > 4294967295`.

### Output for `myvar = D`:

```
myvar: 2147483648
myvar > INT_MAX: True
myvar == INT_MAX: False
myvar > UINT_MAX: False
myvar > LONG_MAX: False
```

### Output for `myvar = E`:

```
myvar: 6294967295
myvar > INT_MAX: True
myvar == INT_MAX: False
myvar > UINT_MAX: True
myvar > LONG_MAX: False
```

**Comment by neikeq** [comment]
cc @bojidar-bg

**Comment by neikeq** [comment]
Since this seems to be an editor bug and not GDScript related, I assume the same will apply to C# once #22005 is merged.

**Comment by bojidar-bg** [comment]
Might be related to #20572.

**Comment by neikeq** [comment]
I forgot this happens on latest master, I tested on 5e57beebb1a43c65d949fdfe5081f9c27f181c80. #20572 was fixed a month ago, so I have my doubts.

**Comment by akien-mga** [comment]
Similar to #26113, I can have a look.

**Comment by akien-mga** [comment]
Some of the issues in the original report are fixed by #26113 already (ints in GDScript no longer overflow until 2^63-1), but the issue with assigning enum values persists. I think it might be the same issue as #26116.

Here are the changes I've attempted so far, but they don't fix the case where you set the exported value from the inspector:
```diff
diff --git a/core/object.cpp b/core/object.cpp
index 4e0416ccb..6fe07ba31 100644
--- a/core/object.cpp
+++ b/core/object.cpp
@@ -79,7 +79,7 @@ PropertyInfo PropertyInfo::from_dict(const Dictionary &p_dict) {
 	PropertyInfo pi;
 
 	if (p_dict.has("type"))
-		pi.type = Variant::Type(int(p_dict["type"]));
+		pi.type = Variant::Type(p_dict["type"].operator int());
 
 	if (p_dict.has("name"))
 		pi.name = p_dict["name"];
@@ -88,7 +88,7 @@ PropertyInfo PropertyInfo::from_dict(const Dictionary &p_dict) {
 		pi.class_name = p_dict["class_name"];
 
 	if (p_dict.has("hint"))
-		pi.hint = PropertyHint(int(p_dict["hint"]));
+		pi.hint = PropertyHint(p_dict["hint"].operator int64_t());
 
 	if (p_dict.has("hint_string"))
 
diff --git a/core/project_settings.cpp b/core/project_settings.cpp
index 6b4895d68..fe18f89bd 100644
--- a/core/project_settings.cpp
+++ b/core/project_settings.cpp
@@ -907,7 +907,7 @@ void ProjectSettings::_add_property_info_bind(const Dictionary &p_info) {
 	ERR_FAIL_INDEX(pinfo.type, Variant::VARIANT_MAX);
 
 	if (p_info.has("hint"))
-		pinfo.hint = PropertyHint(p_info["hint"].operator int());
+		pinfo.hint = PropertyHint(p_info["hint"].operator int64_t());
 	if (p_info.has("hint_string"))
 		pinfo.hint_string = p_info["hint_string"];
 
diff --git a/editor/editor_properties.cpp b/editor/editor_properties.cpp
index 98950023c..6c26a9219 100644
--- a/editor/editor_properties.cpp
+++ b/editor/editor_properties.cpp
@@ -801,11 +801,11 @@ EditorPropertyLayers::EditorPropertyLayers() {
 void EditorPropertyInteger::_value_changed(double val) {
 	if (setting)
 		return;
-	emit_changed(get_edited_property(), int(val));
+	emit_changed(get_edited_property(), (int64_t)val);
 }
 
 void EditorPropertyInteger::update_property() {
-	int val = get_edited_object()->get(get_edited_property());
+	int64_t val = get_edited_object()->get(get_edited_property());
 	setting = true;
 	spin->set_value(val);
 	setting = false;
diff --git a/editor/editor_settings.cpp b/editor/editor_settings.cpp
index cbd8df315..971f4dba5 100644
--- a/editor/editor_settings.cpp
+++ b/editor/editor_settings.cpp
@@ -235,7 +235,7 @@ void EditorSettings::_add_property_info_bind(const Dictionary &p_info) {
 	ERR_FAIL_INDEX(pinfo.type, Variant::VARIANT_MAX);
 
 	if (p_info.has("hint"))
-		pinfo.hint = PropertyHint(p_info["hint"].operator int());
+		pinfo.hint = PropertyHint(p_info["hint"].operator int64_t());
 	if (p_info.has("hint_string"))
 		pinfo.hint_string = p_info["hint_string"];
 
diff --git a/modules/gdnative/nativescript/nativescript.cpp b/modules/gdnative/nativescript/nativescript.cpp
index 9fd0a2e8e..859daa916 100644
--- a/modules/gdnative/nativescript/nativescript.cpp
+++ b/modules/gdnative/nativescript/nativescript.cpp
@@ -672,7 +672,7 @@ void NativeScriptInstance::get_property_list(List<PropertyInfo> *p_properties) c
 
 				PropertyInfo info;
 
-				info.type = Variant::Type(d["type"].operator int64_t());
+				info.type = Variant::Type(d["type"].operator int());
 				ERR_CONTINUE(info.type < 0 || info.type >= Variant::VARIANT_MAX);
 
 				info.name = d["name"];
diff --git a/modules/gdscript/gdscript.cpp b/modules/gdscript/gdscript.cpp
index dd9b36fd8..5cc203a12 100644
--- a/modules/gdscript/gdscript.cpp
+++ b/modules/gdscript/gdscript.cpp
@@ -1094,7 +1094,7 @@ void GDScriptInstance::get_property_list(List<PropertyInfo> *p_properties) const
 					pinfo.name = d["name"];
 					ERR_CONTINUE(pinfo.name == "");
 					if (d.has("hint"))
-						pinfo.hint = PropertyHint(d["hint"].operator int());
+						pinfo.hint = PropertyHint(d["hint"].operator int64_t());
 					if (d.has("hint_string"))
 						pinfo.hint_string = d["hint_string"];
 					if (d.has("usage"))
diff --git a/modules/gdscript/gdscript_parser.cpp b/modules/gdscript/gdscript_parser.cpp
index a642abeb0..4341de7eb 100644
--- a/modules/gdscript/gdscript_parser.cpp
+++ b/modules/gdscript/gdscript_parser.cpp
@@ -4879,7 +4879,7 @@ void GDScriptParser::_parse_class(ClassNode *p_class) {
 			case GDScriptTokenizer::TK_PR_ENUM: {
 				//multiple constant declarations..
 
-				int last_assign = -1; // Incremented by 1 right before the assignment.
+				int64_t last_assign = -1; // Incremented by 1 right before the assignment.
 				String enum_name;
 				Dictionary enum_dict;
 
diff --git a/modules/visual_script/visual_script.cpp b/modules/visual_script/visual_script.cpp
index 2c89e5a35..db2cb5f8f 100644
--- a/modules/visual_script/visual_script.cpp
+++ b/modules/visual_script/visual_script.cpp
@@ -685,11 +685,11 @@ void VisualScript::_set_variable_info(const StringName &p_name, const Dictionary
 
 	PropertyInfo pinfo;
 	if (p_info.has("type"))
-		pinfo.type = Variant::Type(int(p_info["type"]));
+		pinfo.type = Variant::Type(p_info["type"].operator int());
 	if (p_info.has("name"))
 		pinfo.name = p_info["name"];
 	if (p_info.has("hint"))
-		pinfo.hint = PropertyHint(int(p_info["hint"]));
+		pinfo.hint = PropertyHint(p_info["hint"].operator int64_t());
 	if (p_info.has("hint_string"))
 		pinfo.hint_string = p_info["hint_string"];
 	if (p_info.has("usage"))
```
